Include/Rover/TrialIterator.hpp

```diff
@@ -0,0 +1,520 @@
+#ifndef ROVER_TRIAL_ITERATOR_HPP
+#define ROVER_TRIAL_ITERATOR_HPP
+#include <iterator>
+#include <optional>
+#include <type_traits>
+
+namespace Rover {
+
+  //! Iterator over a Trial.
+  template<typename, typename = void>
+  class TrialIterator;
+
+  //! Type traits to check whether a Trial returns Samples by copy.
+  template<typename T>
+  inline constexpr bool returns_sample_by_copy_v = std::is_same_v<decltype(
+    std::declval<T>()[std::declval<std::size_t>()]), typename T::Sample>;
+
+  //! Type traits to check whether a Trial returns Samples by reference.
+  template<typename T>
+  inline constexpr bool returns_sample_by_reference_v = std::is_same_v<
```

Comment:
I would prefer the following check ```std::is_reference_v<decltype(...)>``` as it minimizes the number of assumptions that are being made about ```operator []``` down to strictly whether or not the ```Trial``` returns a reference. Then your ```returns_sample_by_copy_v``` is the negation of ```returns_sample_by_reference_v```.